mmaprototype: pass []roachpb.StoreID for computeMeansForStoreSet

wenyihu6 opened the pull request:
Epic: CRDB-55052 
Release note: none

---
**mmaprototype: pass []roachpb.StoreID for computeMeansForStoreSet**

Previously, computeMeansForStoreSet took in a meansForStoreSet struct. Future
PRs will use this helper function to compute mean load summaries for stores
without having a storeIDPostingList handy. To avoid unnecessary construction of
a storeIDPostingList, the function signature now takes meansLoad and
[]roachpb.StoreID directly. This change lets future callers pass a simple slice
of stores. Since future callers may provide slices with duplicate store IDs, the
function now also de-duplicates them internally.

---
**mmaprototype: refactor computeMeansForStoreSet**

Previously, we updated computeMeansForStoreSet to take meansLoad and
[]roachpb.StoreID directly instead of a meansForStoreSet struct. This commit
updates the call sites that only require meansLoad to use it directly, removing
unnecessary use of meansForStoreSet.

---
**mmaprototype: pass in scratch stores for computeMeansForStoreSet**

Previously, computeMeansForStoreSet allocated a new map on every call to
deduplicate the provided stores list. This commit refactors it to reuse a
scratchStores map (similar to scratchNodes). The caller now allocates this map
once and stores it in the function scope or struct, reducing repeated
allocations.

---
**mmaprototype: return meansLoad directly for computeMeansForStoreSet**

Previously, computeMeansForStoreSet received *meansForStoreSet, which contained
information like stores needed to compute the mean. A recent commit changed it
to take *meansLoad, and computeMeansForStoreSet doesn’t actually use any fields
as provided information. This commit updates computeMeansForStoreSet to directly
construct and return a meanLoad instead.

wenyihu6 committed:
mmaprototype: pass []roachpb.StoreID for computeMeansForStoreSet

Previously, computeMeansForStoreSet took in a meansForStoreSet struct. Future
PRs will use this helper function to compute mean load summaries for stores
without having a storeIDPostingList handy. To avoid unnecessary construction of
a storeIDPostingList, the function signature now takes meansLoad and
[]roachpb.StoreID directly. This change lets future callers pass a simple slice
of stores. Since future callers may provide slices with duplicate store IDs, the
function now also de-duplicates them internally.

wenyihu6 committed:
mmaprototype: pass in scratch stores for computeMeansForStoreSet

Previously, computeMeansForStoreSet allocated a new map on every call to
deduplicate the provided stores list. This commit refactors it to reuse a
scratchStores map (similar to scratchNodes). The caller now allocates this map
once and stores it in the function scope or struct, reducing repeated
allocations.

wenyihu6 committed:
mmaprototype: refactor computeMeansForStoreSet

Previously, we updated computeMeansForStoreSet to take meansLoad and
[]roachpb.StoreID directly instead of a meansForStoreSet struct. This commit
updates the call sites that only require meansLoad to use it directly, removing
unnecessary use of meansForStoreSet.

tbg reviewed (approved):
LGTM, thanks!

> @tbg reviewed 2 of 2 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154688#-:-OaeEsXI8Pe592yTqBJz:b2c0zcb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/load.go` line 375 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/154688#-OaeEAYF-K8GRo3tnhNx:-OaeEAYF-K8GRo3tnhNy:b51qhy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/49e5b1c2d5223ce041a8fe37692d7f194f5ae998/pkg/kv/kvserver/allocator/mmaprototype/load.go#L375)):*
> ```Go
> func computeMeansForStoreSet(
> 	loadProvider loadInfoProvider,
> 	means *meansLoad,
> ```

At this point, wouldn't it be more idiomatic to return a `meansLoad` from this method (vs. receiving one and then mutating it)?


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
mmaprototype: return meansLoad directly for computeMeansForStoreSet

Previously, computeMeansForStoreSet received *meansForStoreSet, which contained
information like stores needed to compute the mean. A recent commit changed it
to take *meansLoad, and computeMeansForStoreSet doesn’t actually use any fields
as provided information. This commit updates computeMeansForStoreSet to directly
construct and return a meanLoad instead.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154688)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/load.go` line 375 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/154688#-OaeEAYF-K8GRo3tnhNx:-Oafyecr3GJhMYGPbcKx:b-lox5ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/49e5b1c2d5223ce041a8fe37692d7f194f5ae998/pkg/kv/kvserver/allocator/mmaprototype/load.go#L375)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

At this point, wouldn't it be more idiomatic to return a `meansLoad` from this method (vs. receiving one and then mutating it)?
</blockquote></details>

Pushed a commit to address this comment. I’ll cherry-pick these commits to my other PR and then rebase that PR on top of this one before merging both.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> @tbg reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154688)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

tbg commented:
CI failure is unrelated: https://github.com/cockroachdb/cockroach/issues/154648

bors r+

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154688)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/allocator/mmaprototype/load.go` line 370 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/154688#-OavFq6QAkwCKoVrrfjt:-OavFq6R2-GdVVk6PWjN:b-oq4bcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68a7d8836ea3e5ecfbcec3224f10469e0ece38b/pkg/kv/kvserver/allocator/mmaprototype/load.go#L370)):*
> ```Go
> // is up to the caller to adjust means.stores if it wants to do filtering.
> //
> // stores may contain duplicate storeIDs, in which case computeMeansForStoreSet
> ```

btw, which caller can provide duplicate storeIDs?


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
@sumeerbhola This is subject to change. More discussion in https://reviewable.io/reviews/cockroachdb/cockroach/154499#-OapuYRJ4hSvRKgmXSYG. Lease queue might pass in duplicate store IDs.
